### PR TITLE
vendor_*.d: include /usr/local by default as "extra" directory

### LINKED ISSUE #5029: Fish doesn't have /usr/local/share/fish/completions in $fish_complete_path
```
$ fish --version
fish, version 2.7.1
```

```
$ uname -a
Linux thinkpad 4.14.47-1-lts #1 SMP Wed May 30 21:09:23 UTC 2018 x86_64 GNU/Linux
```

```
$ echo $fish_complete_path
/home/blevy/.config/fish/completions /etc/fish/completions /usr/share/fish/vendor_completions.d /usr/share/fish/completions /home/blevy/.local/share/fish/generated_completions
```

I downloaded `fish` through the default arch repos.

### PR DESCRIPTION
**Pull request by zanchey**
This PR builds on the work in #6428, but extends it to update the documentation and actually include `/usr/local` by default as the "extra" configuration directory.

Closes #5029.

### CODE REVIEW

**Comment by faho** [comment]
Okay, so what this does is add /usr/local/share as the one directory that's always included *instead* of $__fish_data_dir.

So if $XDG_DATA_DIRS is set, but doesn't include $__fish_data_dir, we'd now use /usr/local/share/fish instead of (most likely) /usr/share/fish.

If we didn't use this, the only way /usr/local/share/fish would be included is if the user or system explicitly added it (e.g. by using /usr/local as the prefix).

It seems weird to me that the prefixed path isn't always in. Is that how these things are usually done?

----

Also @krobelus: While trying to understand this I noticed that the path without $XDG_DATA_DIRS probably doesn't work:

```fish
if set -q XDG_DATA_DIRS
    set --path xdg_data_dirs $XDG_DATA_DIRS
    set xdg_data_dirs (string replace -r '([^/])/$' '$1' -- $xdg_data_dirs)
else
    set xdg_data_dirs $__fish_data_dir
end
```

$__fish_data_dir already includes the "/fish" part, while $XDG_DATA_DIRS won't. So since the uses later add "/fish", it can't work with $__fish_data_dirs.

Committing a fix in a moment.

(and that `string replace` is probably unnecessary - it removes trailing slashes, but double slashes are allowed in paths)

**Comment by krobelus** [comment]
My bad, I completely forgot that we ought to include `/usr/local` independently of `PREFIX` by default.
Truth is, I'm not really familiar with systems that don't define `$XDG_DATA_DIRS`. Those who do most likely already include `/usr/local`.

The problem with this change is that it would reintroduce #5719 because the extra directories are created.

As I understand from reading 484c1484c9ae01354b1910bae5ed8a729e00b3be the reason for having extra directories is to support packaging systems like Homebrew and Nix, that probably install each package to its own prefix (?).
I don't know who else uses the pkgconfig interface, and thus the implicitations of changing the default from `PREFIX/share` to `/usr/local/share`. Packagers might have hardcoded `PREFIX/share/vendor_completions.d`.

Maybe we can create and include `$PREFIX/share/fish/vendor_completions.d` and include but not create `$extra_completions`?


---

> Okay, so what this does is add /usr/local/share as the one directory that's always included _instead_ of $__fish_data_dir.
> 
> So if $XDG_DATA_DIRS is set, but doesn't include $__fish_data_dir, we'd now use /usr/local/share/fish instead of (most likely) /usr/share/fish.

It's probably reasonable to include `PREFIX/share/fish` as well. As I said I'd have to look into systems where this applies (e.g. no `XDG_DATA_DIRS`).

> Also @krobelus: While trying to understand this I noticed that the path without $XDG_DATA_DIRS probably doesn't work:

Whoops, thanks!

> (and that `string replace` is probably unnecessary - it removes trailing slashes, but double slashes are allowed in paths)

`string replace` here makes sure the path deduplication works, because `$XDG_DATA_DIRS` is usually defined with trailing slashes

**Comment by zanchey** [comment]
> Okay, so what this does is add /usr/local/share as the one directory that's always included _instead_ of $__fish_data_dir.

No, I don't think that's the case?

https://github.com/fish-shell/fish-shell/blob/a7ed9e04b369a2107939499831297ca4ea4451cc/share/config.fish#L75-L81 

This should ensure that the datadir is always included.

> Maybe we can create and include `$PREFIX/share/fish/vendor_completions.d` and include but not create `$extra_completions`

Yes, that seems reasonable.

> Packagers might have hardcoded `PREFIX/share/vendor_completions.d`.

The reason for supplying a pkg-config file is to avoid this and allow configure and friends to pick it up at build time.

The problem is there's actually no easy right answer here for developers of other packages - the best we can do is say "use pkg-config, but allow it to be overridden by (eg) dpkg-buildpackage/rpmbuild", as when built by package managers they should install to the prefix, but outside of that should generally go in `/usr/local` .

**Comment by zanchey** [comment]
> Okay, so what this does is add /usr/local/share as the one directory that's always included _instead_ of $__fish_data_dir.

Oh, sorry, I'm with you now.

If `XDG_DATA_DIRS` are defined, I think we should drop everything else.

**Comment by zanchey** [comment]
See also https://github.com/Homebrew/homebrew-core/pull/2813 where I changed the installation of the "outside of the prefix" directories to soft-fail, which I think is fine for #5719, though @SanskritFritz might like to comment?

File: sphinx_doc_src/index.rst [hunk 1]
```diff
@@ -383,7 +383,7 @@ By default ``$fish_function_path`` contains the following:
 
 - A directory for end-users to keep their own functions, usually ``~/.config/fish/functions`` (controlled by the ``XDG_CONFIG_HOME`` environment variable).
 - A directory for systems administrators to install functions for all users on the system, usually ``/etc/fish/functions`` (really ``$__fish_sysconfdir/functions``).
-- A directory for third-party software vendors to ship their own functions for their software, usually ``/usr/share/fish/vendor_functions.d``. (set at compile time; by default, ``$__fish_data_dir/vendor_functions.d``)
+- Directories for third-party software vendors to ship their own functions for their software. Fish searches the directories in the ``XDG_DATA_DIRS`` environment variable for a ``share/fish/vendor_functions.d`` directory; if this variable is not defined, the default is usually to search ``/usr/share/fish/vendor_functions.d`` and ``/usr/local/share/fish/vendor_functions.d``.
```
Comment (faho) [inline sphinx_doc_src/index.rst]: Just "fish/vendor*", not "share/fish/vendor" - the "share" is already in $XDG_DATA_DIRS.

File: cmake/Install.cmake [hunk 2]
```diff
@@ -113,6 +113,9 @@ INSTALL(FILES share/config.fish
 # -$v $(INSTALL) -m 755 -d $(DESTDIR)$(extra_completionsdir)
 # -$v $(INSTALL) -m 755 -d $(DESTDIR)$(extra_functionsdir)
 # -$v $(INSTALL) -m 755 -d $(DESTDIR)$(extra_confdir)
```
Comment (faho) [inline cmake/Install.cmake]: I think it's time to remove these comments. Doesn't have to be in this PR, but I think we're truly past make now.

**Comment by krobelus** [comment]
> See also [Homebrew/homebrew-core#2813](https://github.com/Homebrew/homebrew-core/pull/2813) where I changed the installation of the "outside of the prefix" directories to soft-fail, which I think is fine for #5719, though @SanskritFritz might like to comment?

The installation of the extra directories still succeeds because the permissions for `/usr/local` and `/usr` are the same.

File: sphinx_doc_src/index.rst [hunk 3]
```diff
@@ -1340,17 +1340,17 @@ By default, Fish searches the following for completions, using the first availab
 
 - A directory for end-users to keep their own completions, usually ``~/.config/fish/completions`` (controlled by the ``XDG_CONFIG_HOME`` environment variable);
 - A directory for systems administrators to install completions for all users on the system, usually ``/etc/fish/completions``;
-- A directory for third-party software vendors to ship their own completions for their software, usually ``/usr/share/fish/vendor_completions.d``;
+- Directories for third-party software vendors to ship their own completions for their software. Fish searches the directories in the ``XDG_DATA_DIRS`` environment variable for a ``share/fish/vendor_completions.d`` directory; if this variable is not defined, the default is usually to search ``/usr/share/fish/vendor_completions.d`` and ``/usr/local/share/fish/vendor_completions.d``;
```
[the remark this answers is not in the document]
Comment (SanskritFritz) [reply sphinx_doc_src/index.rst]: I built fish and all directories are perfectly at their place. Thanks!

**Comment by zanchey** [comment]
I'm away from anything with a proper keyboard for a few days, but I'll get this merged and get 3.1b1 out the door once I get back.

File: sphinx_doc_src/index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (krobelus) [reply sphinx_doc_src/index.rst]: It still creates the extra directories though
```
$ git fetch
$ git checkout refs/pullreqs/6508
$ rm -rf build
$ mkdir build
$ cd build
$ cmake .. -DCMAKE_INSTALL_PREFIX=/tmp/fish -DBUILD_DOCS=OFF -DCMAKE_BUILD_TYPE=Debug
$ make -j
$ ls /usr/local/share/fish
ls: cannot access '/usr/local/share/fish': No such file or directory
$ sudo make install
$ ls /usr/local/share/fish
vendor_completions.d/  vendor_conf.d/  vendor_functions.d/
```
I think it's better to let packagers create these directories themselves if they need it.

**Comment by zanchey** [comment]
There's really no right answer unfortunately. I'm happy to examine what zsh does, and follow that.

**Comment by zanchey** [comment]
OK, let's make fish do what zsh does - allow an extra directory to be added to the search path, but not create it (and leave that for downsteam). fish will create a vendor directory inside its prefix, and Homebrew already creates the extra directory, so I think everything will continue to work.

**Pull request closed without merging**